idea-plugin/src/main/kotlin/io/github/composegears/valkyrie/ui/screen/mode/iconpack/existingpack/ui/viewmodel/ExistingPackViewModel.kt

```diff
@@ -0,0 +1,162 @@
+
+    fun onAction(existingPackAction: ExistingPackAction) {
+        when (existingPackAction) {
+            is SelectKotlinFile -> onChooseFile(existingPackAction.path)
+            is AddNestedPack -> inputHandler.addNestedPack()
+            is RemoveNestedPack -> inputHandler.removeNestedPack(existingPackAction.nestedPack)
+            is PreviewPackObject -> previewIconPackObject()
+            is SavePack -> saveIconPack()
+        }
+    }
+
+    private fun onChooseFile(path: Path) {
+        val iconPackInfo = IconPackPsiParser.extractIconPack(path) ?: return
+
+        val inputFieldState = iconPackInfo.toInputFieldState()
+        inputHandler.updateState(inputFieldState)
+
+        _state.updateState {
+            ExistingPackEditState(
+                packEditState = PackEditState(inputFieldState = inputFieldState),
+                exportDirectory = path.parent.absolutePathString(),
+            )
+        }
+    }
+
+    private fun previewIconPackObject() = viewModelScope.launch {
+        val editState = currentState.castOrNull<ExistingPackEditState>()?.packEditState ?: return@launch
+        val inputFieldState = editState.inputFieldState
+
+        val iconPackCode = IconPackGenerator.create(
+            config = IconPackGeneratorConfig(
+                packageName = inputFieldState.packageName.text,
+                iconPackName = inputFieldState.iconPackName.text,
+                subPacks = inputFieldState.nestedPacks.map { it.inputFieldState.text },
+            ),
+        ).content
+
+        _events.emit(PreviewIconPackObject(code = iconPackCode))
+    }
+
+    private fun saveIconPack() {
+        val editState = currentState.castOrNull<ExistingPackEditState>() ?: return
+        val inputFieldState = editState.packEditState.inputFieldState
+
+        viewModelScope.launch {
+            inMemorySettings.updatePackageName(inputFieldState.packageName.text)
+            inMemorySettings.updateIconPackName(inputFieldState.iconPackName.text)
+            inMemorySettings.updateNestedPack(inputFieldState.nestedPacks.map { it.inputFieldState.text })
+            inMemorySettings.updateMode(Mode.IconPack)
+            inMemorySettings.updateIconPackDestination(editState.exportDirectory)
+
+            val iconPack = IconPackGenerator.create(
+                config = IconPackGeneratorConfig(
+                    packageName = inMemorySettings.current.packageName,
+                    iconPackName = inMemorySettings.current.iconPackName,
+                    subPacks = inMemorySettings.current.nestedPacks,
+                ),
+            )
+
+            FileWriter.writeToFile(
+                content = iconPack.content,
+                outDirectory = inMemorySettings.current.iconPackDestination,
+                fileName = iconPack.name,
+            )
+
+            writeAction {
+                VirtualFileManager.getInstance().syncRefresh()
+            }
+            _events.emit(OnSettingsUpdated)
+        }
```

Comment:
Duplicate with `NewPackViewModel#saveIconPack`?

Reply:
difference in single line, will reuse